### ISSUE TITLE
Timezones not supported when using Waterline standalone.

### ISSUE THREAD

**Issue by eddieajau** [balderdashy/sails-mysql#144]
I'm having trouble storing dates in UTC using Waterline standalone with this adapter. I think I've traced the problem to here:

https://github.com/balderdashy/sails-mysql/blob/master/lib/adapter.js#L383

Before that, a `datetime` is a Date object, after it is a string.

However, `utils.toSqlDate` (https://github.com/balderdashy/sails-mysql/blob/master/lib/utils.js#L249) does not seem to make any attempt to adjust the timezone.

My Waterline configuration is:

```js
var mysqlAdapter = require('sails-mysql');

module.exports = function(config) {
	var mysql = config.get('mysql');

	return {
		adapters: {
			mysql: mysqlAdapter
		},

		connections: {
			default: {
				adapter: 'mysql',
				host: mysql.host || '127.0.0.1',
				port: mysql.port || 3306,
				user: mysql.user,
				password: mysql.password,
				database: mysql.database,
				timezone: '+0'
			}
		},

		defaults: {
			migrate: 'safe',
			autoCreatedAt: false,
			autoUpdatedAt: false
		}
	};
};
```

My models are set up like this:

```js
		created_ts: {
			type: 'datetime',
			defaultsTo: function () {
				return new Date();
			}
		},
		updated_ts: {
			type: 'datetime'
		},
```

I know the DB is set to UTC because I have an `ON UPDATE CURRENT_TIMESTAMP` on the `updated_ts` field. After saving the model, the `created_ts` is in local time, and the `updated_ts` is in UTC.

**Comment by mikeharty** [comment]
When migration mode is set to alter, waterline is pulling all of the records out of the DB and re-creating them, translating them to UTC from the system timezone, but they have already gone through this translation, so they gain/lose the difference every time.

**Comment by mikeharty** [comment]
A simple solution to this may be to set a semaphore on the sails object like "currentlySyncing" before the createEach function is called from the alter strategy, check this semaphore before translating the date in the adapter, and the turn the flag off after everything is re-inserted.

**Comment by mikeharty** [comment]
I realize this is not generic because Waterline can be used independently of Sails, but for reference, this solves the issue for me. If you can give me a suggestion about where this semaphore would make more sense to be set or passed, I'll happily send a PR.

**/node_modules/sails/node_modules/waterline/lib/waterline/adapter/sync/strategies/alter.js @ 186**
```javascript
sails.currentlySyncing = true;
self.createEach(backupData, function(err) {
  if (err) return uhoh(err, backupData, cb);

  // Done.
  sails.currentlySyncing = false;
  return cb();
});
```
**/node_modules/sails-mysql/lib/utils.js @ 249**
```javascript
utils.toSqlDate = function toSqlDate(date) {
    if(typeof(sails.currentlySeeking) !== 'undefined' && !sails.currentlySyncing) {
        date = date.getUTCFullYear() + '-' +
        ('00' + (date.getUTCMonth() + 1)).slice(-2) + '-' +
        ('00' + date.getUTCDate()).slice(-2) + ' ' +
        ('00' + date.getUTCHours()).slice(-2) + ':' +
        ('00' + date.getUTCMinutes()).slice(-2) + ':' +
        ('00' + date.getUTCSeconds()).slice(-2);
    }
  return date;
};
```

**Comment by mikeharty** [comment]
Realized this is not a complete solution. I only had the described behavior when setting the timezone on my connection to Zulu, so I changed the toSqlDate to use UTC and removed the connection timezone. This works if you actually want UTC, but if you specify a different timezone in your connection you will still have a similar problem.

I'll dig into this some more when I have time.

**Comment by vincesp** [comment]
This also seems to break the handling of fields that are defined as `date` (vs. `datetime`). Posting the following JSON:

    {
      "firstName": "John",
      "lastName": "Smith",
      "birthDate": "2010-10-10"
    }

returns

    {
      "firstName": "John",
      "lastName": "Smith",
      "birthDate": "2010-10-09T04:00:00.000Z",
      "id": 1,
      "createdAt": "2015-04-13T17:08:24.000Z",
      "updatedAt": "2015-04-13T17:08:24.000Z"
    }

The database column contains `2010-10-09` without a time.
I suggest that `date` and `datetime` columns are treated separately.

**Comment by vincesp** [comment]
Performing some more tests: this behaviour appeared on an openshift.com instance. Deploying the same app to heroku.com, I am able to produce the expected result.

    {
      "firstName": "James",
      "lastName": "Joyce",
      "birthDate": "2010-10-10T00:00:00.000Z",
      "id": 2,
      "createdAt": "2015-04-13T15:56:34.000Z",
      "updatedAt": "2015-04-13T16:01:35.000Z"
    }

**Comment by devinivy** [comment]
Sounds like we need to write some tests to get passing?

**Comment by eddieajau** [comment]
So, the **mysql** package has a static utility class called `SqlString`, and it has a method `dateToString(date, timeZone)`. I think it would be good to use this in place of the `toSqlDate` function in `sails-mysql/lib/sql.js` and `utils.js` which is called in `prepareValue`.

I did a rough test with the following code:

```js
var SqlString = require('sails-mysql/node_modules/mysql/lib/protocol/SqlString');
utils.toSqlDate = function toSqlDate(date) {
	return SqlString.dateToString(date, 'Z');
};
```

and that stored my date correctly in UTC. The trick is getting the timezone configuration in there.

Am I on the right track to solving this or is there a better _Sails Way_?

Thanks in advance.

**Comment by sfakir** [comment]
+1
I support @eddieajau  idea to just use the methods of node-mysql

**Comment by sailsbot** [comment]
Thanks for posting, @eddieajau.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
Thanks for posting, @eddieajau.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by wassimsalib** [comment]
Hi Everyone, any updates on this issue? I have a model with data type "date" and whenever the client sends a full datetime string (i.e. "2016-01-15T06:00:00Z") sometimes it is saved on the DB with -1 "2016-01-14", but what I have noticed that this problem doesn't always happen. it only happens when the UTC time is in a different day (past 12:00 am) than my local time (Mountain time -6).

**Comment by tomercagan** [comment]
@wassimsalib I have run into this as well. With date, you send a date time (js) but then it is saved in the DB local time. In case you're close to the end of the day (e.g. 00:00:00) and the DB is in different timezone, it will convert it to the local representation. For example, "2016-01-15T06:00:00Z" becomes "2016-01-14T23:00:00Z" (assuming local TZ is -7) and that is then truncated because you're only saving the date part. Try to stick to datetime to avoid this.
